File: structy/structy-py/3-Binary-Tree/treeValueCount.py
# Write a function, tree_value_count, that takes in the root of a binary tree and a target value. The function should return the number of times that the target occurs in the tree.

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


# def tree_value_count(root, target):
#     count = 0
#     stack = [root]
#     if root is None:
#         return 0
#     while stack:
#         node = stack.pop()
#         if node.val == target:
#             count += 1
#         if node.left:
#             stack.append(node.left)
#         if node.right:
#             stack.append(node.right)
#     return count

#! recursion


def tree_value_count(root, target):
    if root is None:
        return 0
    count = 1 if root.val == target else 0
    left = tree_value_count(root.left, target)
    right = tree_value_count(root.right, target)
    return count + left + right


#!TEST_00
a = Node(12)
b = Node(6)
c = Node(6)
d = Node(4)
e = Node(6)
f = Node(12)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      12
#    /   \
#   6     6
#  / \     \
# 4   6     12

print(tree_value_count(a,  6))  # -> 3


#!TEST_01
a = Node(12)
b = Node(6)
c = Node(6)
d = Node(4)
e = Node(6)
f = Node(12)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      12
#    /   \
#   6     6
#  / \     \
# 4  6     12

print(tree_value_count(a,  12))  # -> 2

#!TEST_02
a = Node(7)
b = Node(5)
c = Node(1)
d = Node(1)
e = Node(8)
f = Node(7)
g = Node(1)
h = Node(1)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g
f.right = h

#      7
#    /   \
#   5     1
#  / \     \
# 1   8     7
#    /       \
#   1         1

print(tree_value_count(a, 1))  # -> 4

#!TEST_03
a = Node(7)
b = Node(5)
c = Node(1)
d = Node(1)
e = Node(8)
f = Node(7)
g = Node(1)
h = Node(1)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g
f.right = h

#      7
#    /   \
#   5     1
#  / \     \
# 1   8     7
#    /       \
#   1         1

print(tree_value_count(a, 9))  # -> 0

#!TEST_04
print(tree_value_count(None, 42))  # -> 0

File: structy/structy-js/Binary-Tree/treeValueCount.js
// Write a function, treeValueCount, that takes in the root of a binary tree and a target value. The function should return the number of times that the target occurs in the tree.

class Node {
	constructor(val) {
		this.val = val;
		this.left = null;
		this.right = null;
	}
}

const treeValueCount = (root, target) => {
	// edge case - if tree / root is null return 0
	if (root === null) return 0;

	let count = 0;
	let queue = [root];
	// transvers through the tree and count each node with target value.
	while (queue.length > 0) {
		const node = queue.shift();
		// console.log(node.val);
		if (node.val === target) count++;
		if (node.left !== null) queue.push(node.left);
		if (node.right !== null) queue.push(node.right);
	}
	return count;
};

//! TEST_00
const a = new Node(12);
const b = new Node(6);
const c = new Node(6);
const d = new Node(4);
const e = new Node(6);
const f = new Node(12);

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      12
//    /   \
//   6     6
//  / \     \
// 4   6     12

console.log(treeValueCount(a, 6)); // -> 3

//! TEST_01
//   const a = new Node(12);
// const b = new Node(6);
// const c = new Node(6);
// const d = new Node(4);
// const e = new Node(6);
// const f = new Node(12);

// a.left = b;
// a.right = c;
// b.left = d;
// b.right = e;
// c.right = f;

// //      12
// //    /   \
// //   6     6
// //  / \     \
// // 4  6     12

// treeValueCount(a,  12); // -> 2

//! TEST_02
//   const a = new Node(7);
// const b = new Node(5);
// const c = new Node(1);
// const d = new Node(1);
// const e = new Node(8);
// const f = new Node(7);
// const g = new Node(1);
// const h = new Node(1);

// a.left = b;
// a.right = c;
// b.left = d;
// b.right = e;
// c.right = f;
// e.left = g;
// f.right = h;

// //      7
// //    /   \
// //   5     1
// //  / \     \
// // 1   8     7
// //    /       \
// //   1         1
// treeValueCount(a, 1); // -> 4

//! TEST_03
//   const a = new Node(7);
// const b = new Node(5);
// const c = new Node(1);
// const d = new Node(1);
// const e = new Node(8);
// const f = new Node(7);
// const g = new Node(1);
// const h = new Node(1);

// a.left = b;
// a.right = c;
// b.left = d;
// b.right = e;
// c.right = f;
// e.left = g;
// f.right = h;

// //      7
// //    /   \
// //   5     1
// //  / \     \
// // 1   8     7
// //    /       \
// //   1         1

// treeValueCount(a, 9); // -> 0

//! TEST_04
//   treeValueCount(null, 42); // -> 0
